src/test/shell/bazel/bazel_unresolved_symlink_test.sh

```diff
@@ -308,4 +415,122 @@ EOF
   assert_contains "input link is /somewhere/in/my/heart" bazel-bin/a/a.file
 }
 
+function test_symlink_created_from_symlink_action() {
+  mkdir -p a
+  cat > a/a.bzl <<'EOF'
+def _a_impl(ctx):
+    target = ctx.actions.declare_file(ctx.label.name + ".target")
+    ctx.actions.write(
+        output = target,
+        content = "Hello, World!",
+    )
+
+    symlink = ctx.actions.declare_file(ctx.label.name + ".link")
+    ctx.actions.symlink(
+        output = symlink,
+        target_file = target,
+    )
+    return DefaultInfo(files = depset([symlink]))
+
+a = rule(implementation = _a_impl)
+EOF
+
+  cat > a/BUILD <<'EOF'
+load(":a.bzl", "a")
+
+a(name="a")
+EOF
+
+  bazel build //a:a || fail "build failed"
+  assert_contains "Hello, World!" bazel-bin/a/a.link
+}
+
+function test_executable_dangling_symlink() {
+  mkdir -p a
+  cat > a/a.bzl <<'EOF'
+def _a_impl(ctx):
+    symlink = ctx.actions.declare_symlink(ctx.label.name + ".link")
+    ctx.actions.symlink(
+        output = symlink,
+        target_path = "/foo/bar",
+        is_executable = True,
+    )
+    return DefaultInfo(files = depset([symlink]))
+
+a = rule(implementation = _a_impl)
+EOF
+
+  cat > a/BUILD <<'EOF'
+load(":a.bzl", "a")
+
+a(name="a")
+EOF
+
+  bazel build --experimental_allow_unresolved_symlinks //a:a && fail "build succeeded"
+  [[ "$?" == 1 ]] || fail "unexpected exit code"
+}
+
+function test_executable_symlink() {
+  mkdir -p a
+  cat > a/a.bzl <<'EOF'
+def _a_impl(ctx):
+    target = ctx.actions.declare_file(ctx.label.name + ".target")
+    ctx.actions.write(
+        output = target,
+        content = "Hello, World!",
+        is_executable = True,
+    )
+
+    symlink = ctx.actions.declare_file(ctx.label.name + ".link")
+    ctx.actions.symlink(
+        output = symlink,
+        target_file = target,
+        is_executable = True,
+    )
+    return DefaultInfo(files = depset([symlink]))
+
+a = rule(implementation = _a_impl)
+EOF
+
+  cat > a/BUILD <<'EOF'
+load(":a.bzl", "a")
+
+a(name="a")
+EOF
+
+  bazel build //a:a || fail "build failed"
+  assert_contains "Hello, World!" bazel-bin/a/a.link
+}
+
+function test_executable_symlink_to_nonexecutable_file() {
+  mkdir -p a
+  cat > a/a.bzl <<'EOF'
+def _a_impl(ctx):
+    target = ctx.actions.declare_file(ctx.label.name + ".target")
+    ctx.actions.write(
+        output = target,
+        content = "Hello, World!",
+    )
+
+    symlink = ctx.actions.declare_file(ctx.label.name + ".link")
+    ctx.actions.symlink(
+        output = symlink,
+        target_file = target,
+        is_executable = True,
+    )
+    return DefaultInfo(files = depset([symlink]))
+
+a = rule(implementation = _a_impl)
+EOF
+
+  cat > a/BUILD <<'EOF'
+load(":a.bzl", "a")
+
+a(name="a")
+EOF
+
+  # bazel build //a:a && fail "build succeeded"
+  # [[ "$?" == 1 ]] || fail "unexpected exit code"
```

Comment:
I can't quite figure out why this isn't failing.
Looking at [this](https://source.bazel.build/bazel/+/master:src/main/java/com/google/devtools/build/lib/analysis/actions/SymlinkAction.java;l=62;drc=79b4a07d35179f9ae29c4c0103bc2ecfcbe5c96c) comment, it seems like the action should fail.
Am I missing something here, or is that comment out-of date (or invalid for Bazel)?

Reply:
Hehe. That's because Bazel makes every output file executable (which is not how I would like it to be, but it would be quite difficult to change now :( ). Probably this is where:

https://github.com/bazelbuild/bazel/blob/master/src/main/java/com/google/devtools/build/lib/skyframe/ActionMetadataHandler.java#L737

If you want to test this functionality, symlink to a source file instead of one that was built by Bazel itself.